### DISCUSSION TITLE
'unnecessary' diagnostics face weight is rendered as normal instead of ultra-light

### DISCUSSION THREAD

**Discussion by Andre0991** [joaotavora/eglot#858] [Q&A] [answered]
Hi. I'm continuing part of the discussion of https://github.com/joaotavora/eglot/discussions/851, because this is another issue. I'm starting it as a discussion because I'm not sure if it's really an issue.

LSP 'unnecessary' diagnostics faces are not rendered with the proper weight, which is `ultra-light`, in my setup.
The face is rendered as if it had a `normal` weight, so it gets basically equal to code with no errors.

The 'unnecessary' diagnostics happens when an argument function is unused, for example.

Now, it seems that the issue is that not many fonts support `ultra-light`.
In fact, I tried these, and apparently, they don't support it:
- DejaVu Sans Mono
- Source Code Pro
- Hack

Those are their variants, according to my OS font list:

![image](https://user-images.githubusercontent.com/7083061/156423297-11a41a6a-ecfd-4546-a5d7-c4f39751e23a.png)
![image](https://user-images.githubusercontent.com/7083061/156423339-0a915c01-9e6e-4d10-98ed-fdce48059ae0.png)
![image](https://user-images.githubusercontent.com/7083061/156423240-c667b107-d3d8-4dba-9777-a8d0d1819166.png)

---

Steps to reproduce:
<details>
Download the latest `clojure-lsp` and add it to path: 

```
wget https://github.com/clojure-lsp/clojure-lsp/releases/download/2022.02.23-12.12.12/clojure-lsp
```

Clone this repo and a simple Clojure project:

```
git clone https://github.com/joaotavora/eglot.git
git clone https://github.com/clojure/tools.trace.git
```

Start Emacs:

```
$ emacs -Q -l flymake -L <path-to-eglot-root> -l eglot.el <path-to-tools.trace-root>/src/main/clojure/clojure/tools/trace.clj -f flymake-mode -f eglot
```

Choose `clojure-lsp` when `eglot` asks for the server.

In line 74, add an extra argument to the function.

```
[name value foo]
```

Now, `foo`'s face is equal to the other arguments.
</details>

Is anyone facing the same issue?

First of all, I'd like to confirm if this really depends on the font. If someone has a font that works, please let me know.

My current workaround is setting `underline`, making it similar to yellow squiggles.

```
(set-face-attribute 'eglot-diagnostic-tag-unnecessary-face nil :weight 'normal :underline "DarkGoldenrod2")
```

Note: For the sake of comparison, `lsp-mode` changes `foreground` rather than `weight`: https://github.com/emacs-lsp/lsp-mode/blob/2d640a146ad164f87c7796826791d7bcb85a8e7a/lsp-diagnostics.el#L61-L69. It adds yellow squiggles as well.

**Comment by leungbk** [comment]
I'm on Mac and Source Code Pro properly displays `ultra-light` for me, but DejaVu Sans Mono does not (both fonts installed via the Nix package manager). I think you're on Mac as well; how can I install fonts as you did (I don't know macOS very well)?

I don't have any understanding of what factors determine whether things like `ultra-light` work. But it may make sense to change the default face for the `unnecessary-face` in Eglot. Even when it works, it doesn't look as good as what lsp-mode uses.

**Reply by joaotavora** [reply]
What is the face that `lsp-mode` uses? Also notice, and I didn't add this note at the time, that there can usually be many face definitions inside the same face depending on the criteria like availability of a certain face property, number of colors, terminal type, etc.

**Reply by Andre0991** [reply]
Hi. I'm using MacOS as well.

This is my Emacs version:
```
GNU Emacs 28.0.60 (build 1, x86_64-apple-darwin19.6.0, NS appkit-1894.60 Version 10.15.7 (Build 19H2)) of 2021-11-25
```

>  how can I install fonts as you did (I don't know macOS very well)?


I installed all fonts using homebrew:

```
brew install font-source-code-pro
brew install font-dejavu
brew install font-hack
```

I just tested source code pro on non graphical Emacs and the face is unchanged as well.

> What is the face that lsp-mode uses?

It adds `:foreground "gray"`. Not sure if it changes more things, aside from the yellow squiggles.

**Reply by leungbk** [reply]
> It adds `:foreground "gray"`. Not sure if it changes more things, aside from the yellow squiggles.

`:foreground "gray"` seems to work with more fonts, looks a lot nicer than `:weight ultra-light`, and even displays properly on terminal Emacs. Maybe we could make this, or something like it, the new default in the face definition. 

@joaotavora would making this change, or a similar one like `:foreground "light-gray"`, violate the copyright terms of the FSF given that we took this idea from lsp-mode?

**Reply by joaotavora** [reply]
@leungbk no, I don't think so, use of a color is not anyone's intellectual property.  But ask this in emacs-devel@gnu.org, and CC to me, thanks.

**Comment by leungbk** [comment, marked as answer]
@Andre0991 I asked on the mailing list and heard from two people that they believed there would be no problem with simply stealing lsp-mode's setting for use in Eglot.

**Reply by joaotavora** [reply]
@leungbk the word "stealing" is a bit strong here no? :smile:  We're literally just observing some color, and thinking how that color would look nice on our own attire.  It's just fashion! :man_dancing: :woman_dancing: 

Anyway, one of the respondents mentioned that it could use inheritance of, perhaps, some font which has the meaning of "subdued", "low-key", "muted" etc... For example, the face `shadow`, seems to have that meaning.

I pushed a change to `master` inheriting from this.  I ask everyone using an LSP server with this capability to give it a go.

And so we won't be "stealing" anything :-)

**Reply by Andre0991** [reply]
Many thanks, @leungbk and @joaotavora.

Here's the result in my machine, using modus-themes. Looks good.

![image](https://user-images.githubusercontent.com/7083061/156930120-1b6ec7d7-c5b1-44af-9e64-4100e523a21d.png)

![image](https://user-images.githubusercontent.com/7083061/156930194-ade2231e-24a1-45f0-943b-42304333a79b.png)
